x/ecocredit/server/testsuite/tx.go

```diff
@@ -0,0 +1,672 @@
+					Quantity:          "1.0",
+					BidPrice:          &bidPrice1,
+					DisableAutoRetire: true,
+				},
+				{
+					Selection:         &ecocredit.MsgBuy_Order_Selection{Sum: &ecocredit.MsgBuy_Order_Selection_SellOrderId{SellOrderId: sellRes.SellOrderIds[1]}},
+					Quantity:          "1.0",
+					BidPrice:          &bidPrice1,
+					DisableAutoRetire: true,
+				},
+			},
+			expErr:  "",
+			wantErr: false,
+			partial: false,
+		},
+		{
+			name:  "valid request - partial fill",
+			buyer: addr2.String(),
+			orders: []*ecocredit.MsgBuy_Order{
+				{
+					Selection:         &ecocredit.MsgBuy_Order_Selection{Sum: &ecocredit.MsgBuy_Order_Selection_SellOrderId{SellOrderId: sellRes.SellOrderIds[2]}},
+					Quantity:          "0.5",
+					BidPrice:          &bidPrice1,
+					DisableAutoRetire: true,
+				},
+				{
+					Selection:         &ecocredit.MsgBuy_Order_Selection{Sum: &ecocredit.MsgBuy_Order_Selection_SellOrderId{SellOrderId: sellRes.SellOrderIds[3]}},
+					Quantity:          "0.5",
+					BidPrice:          &bidPrice1,
+					DisableAutoRetire: true,
+				},
+			},
+			expErr:  "",
+			wantErr: false,
+			partial: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		s.Run(tc.name, func() {
+			require := s.Require()
+
+			// get buyer coin balance before
+			coinBalanceBefore := s.bankKeeper.GetBalance(s.sdkCtx, addr2, "stake")
+
+			// get buyer credit balance before
+			creditBalanceBefore, _ := s.queryClient.Balance(s.ctx, &ecocredit.QueryBalanceRequest{
+				Account:    addr2.String(),
+				BatchDenom: createBatchRes.BatchDenom,
+			})
+
+			// process buy orders
+			res, err := s.msgClient.Buy(s.ctx, &ecocredit.MsgBuy{
+				Buyer:  tc.buyer,
+				Orders: tc.orders,
+			})
+
+			// get buyer coin balance after
+			coinBalanceAfter := s.bankKeeper.GetBalance(s.sdkCtx, addr2, "stake")
+
+			// get buyer credit balance after
+			creditBalanceAfter, _ := s.queryClient.Balance(s.ctx, &ecocredit.QueryBalanceRequest{
+				Account:    addr2.String(),
+				BatchDenom: createBatchRes.BatchDenom,
+			})
+
+			// query first sell order
+			sellResponse1, sellError1 := s.queryClient.SellOrder(s.ctx, &ecocredit.QuerySellOrderRequest{
+				SellOrderId: tc.orders[0].Selection.GetSellOrderId(),
+			})
+
+			// query second sell order
+			sellResponse2, sellError2 := s.queryClient.SellOrder(s.ctx, &ecocredit.QuerySellOrderRequest{
+				SellOrderId: tc.orders[1].Selection.GetSellOrderId(),
+			})
+
+			if tc.wantErr {
+				require.Error(err)
+				require.Contains(err.Error(), tc.expErr)
+				require.Equal(coinBalanceBefore, coinBalanceAfter)
+				require.Equal(creditBalanceBefore, creditBalanceAfter)
+			} else {
+				require.NoError(err)
+				require.NotNil(res.BuyOrderIds)
+				require.NotEqual(coinBalanceBefore, coinBalanceAfter)
+				require.NotEqual(creditBalanceBefore, creditBalanceAfter)
```

Comment:
would prefer to have exacts checked here to ensure tx math is correct

Reply:
added in https://github.com/regen-network/regen-ledger/pull/626/commits/e7fe6a9e4099aeeceacb344b0ffc401b2192b03f